File: kotlin/dp/main_10844.kt
fun main() {
    val num = readLine()!!.toInt()
    var prev = listOf(0, 1, 1, 1, 1, 1, 1, 1, 1, 1)
    val now = mutableListOf<Int>()
    for (i in 1 until num) {
        now.clear()
        (0..9).forEach {
            now.add(it, when (it) {
                0 -> prev[it + 1]
                9 -> prev[it - 1]
                else -> (prev[it + 1] + prev[it - 1]) % 1000000000
            })
        }
        prev = now.toList()
    }
    println(prev.sumOf { it.toLong() } % 1000000000)
}

File: cpp/step/dp/main_10844.cpp
#include <iostream>

using namespace std;

int main() {
	int num;
	cin >> num;
	
	int check[2][10] = { {0, 1, 1, 1, 1, 1, 1, 1, 1, 1}, {0, } };
	for (int i = 1; i < num; ++i) {
		int prev = (i + 1) % 2;
		int now = i % 2;
		check[now][0] = check[prev][1];
		for (int j = 1; j < 9; ++j) {
			check[now][j] = (check[prev][j - 1] + check[prev][j + 1]) % 1000000000;
		}
		check[now][9] = check[prev][8];
	}
	
	long long ans = 0;
	for (int i = 0; i < 10; ++i) {
		ans = (ans + check[(num + 1) % 2][i]) % 1000000000;
	}
	cout << ans << endl;
}
